src/modules/dna-service/dna.background.ts

```diff
@@ -248,7 +260,56 @@ export class DnaService {
     return browserStorage.set(ACTION_QUEUE_KEY, []);
   }
 
-  initialize() {
+  async registerAllWallets() {
+    // To make sure referral codes are generated for new wallets we need to call 'registerWallet' on the Zerion DNA Service.
+    // Existing "owned" wallets (wallets with provider) require to have referral codes as well,
+    // but the backend lacks sufficient data to assign them.
+    //
+    // Additionally, re-registration is crucial due to a previous backend error:
+    // Zerion DNA Service expected a version prefix 'v' for 'registerWallet' requests, which we were not including.
+    // This backend issue has since been resolved, but as a result,
+    // wallets previously registered without 'v' are effectively unregistered.
+    // To handle this, we have to re-register all existing wallets.
+
+    const isInvoked = await browserStorage.get<boolean>(
+      REGISTER_ALL_WALLETS_INVOKED_KEY
+    );
+    if (isInvoked) {
+      return;
+    }
+
+    const wallet = this.getWallet();
+    const walletGroups = await wallet.uiGetWalletGroups({
+      context: INTERNAL_SYMBOL_CONTEXT,
+    });
+
+    const ownedAddressesWithGroupOrigins =
+      walletGroups
+        ?.filter((group) => !isReadonlyContainer(group.walletContainer))
+        ?.flatMap((group) =>
+          group.walletContainer.wallets.map((wallet) => ({
+            address: wallet.address,
+            origin: group.origin || undefined,
+          }))
+        ) || [];
+
+    const walletRegistrationRequests = ownedAddressesWithGroupOrigins.map(
+      ({ address, origin }) =>
+        this.registerWallet({ params: { address, origin } })
+    );
+
+    const results = await Promise.allSettled(walletRegistrationRequests);
+    const hasRejectedRequests = results.some(
+      (result) => result.status === 'rejected'
+    );
+
+    if (!hasRejectedRequests) {
+      await browserStorage.set(REGISTER_ALL_WALLETS_INVOKED_KEY, true);
+    }
+  }
+
+  initialize({ account }: { account: Account }) {
+    account.on('authenticated', this.registerAllWallets.bind(this));
```

Comment:
Does this mean we will only call registerAllWallets when using logs in? What if the user never logs out?

Reply:
I see, that's a good point